### DISCUSSION TITLE
How to identify inside and outside faces using face normal within pyvista?

### DISCUSSION THREAD

**Discussion by venugovh** [pyvista/pyvista#2639] [Q&A]
Hi! 

I am trying to find the inside and outside facing STL facets based on its outward normal information. To be precise, assuming that the triangular facet coordinates are oriented in the counter-clockwise direction in pyvista, the normal already embedded as `mesh.face_normals` are outward normals, using the right-hand rule.

My plan is to identify each inside and outside facing polygons for one slice of the STL file. In the grand scheme of things, I am trying to sample points inside the polygons whose normals are oriented inside, and I am going to repeat this process over an array of slices. Are there any filters that already exist in pyvista for identifying the inside and outside facing polygon edges from the slice information or the original mesh polydata? I would appreciate it if you can direct me to any libraries that does this task. Thank you!

A sample slice image is shown below.

![image](https://user-images.githubusercontent.com/47824150/131923708-d0a2c728-ac55-4117-ad3c-038c8f3bed52.png)

The sample STL part and the code is shown below. 

```
import pyvista as pv
import numpy as np

# import mesh
mesh = pv.read('test_part.stl')

#uniform scaling (optional)
mesh_bounds = np.asarray(mesh.bounds)
mesh.points/=np.max(mesh_bounds)

# generate one slice of the part
slice = mesh.slice_along_axis(1, axis='z')

# plotter
p = pv.Plotter()
p.set_background('w')
p.add_mesh(mesh, color='r', opacity=0.5)
p.add_mesh(slice, color='k', opacity=1, line_width=5)
p.show()

```

The STL file looks like this. I am not able to attach the STL file here!
![image](https://user-images.githubusercontent.com/47824150/131924089-4907b362-0a2b-45e6-839d-721de81abdee.png)

**Comment by adam-grant-hendry** [comment]
@Vysakh5391 For 2D data, I would use the `shapely` library to segment your data. It can provide a hierarchy of exterior and interior polygons, similar to the `opencv` method `connectedComponentsWithStats`:

(For details, see 
    [YouTube: Python shapely tutorial part 1](https://www.youtube.com/watch?v=LwpqA2WMR_8) from `10:11` to end, and
    [GitHub: Corresponding code](https://github.com/mmccoo/nerd_mmccoo/blob/master/shapely_tutorial/polygon.py))

Alternatively, for 3D data, I would recommend exploring alpha shapes (see [Wikipedia](https://en.wikipedia.org/wiki/Alpha_shape) and [Introduction to Alpha Shapes](https://graphics.stanford.edu/courses/cs268-11-spring/handouts/AlphaShapes/as_fisher.pdf)). The [open3d library](http://www.open3d.org/docs/latest/tutorial/geometry/surface_reconstruction.html) provides a great set of tools for alpha shapes, ball pivoting, and poisson reconstruction surface reconstruction algorithms. In addition, there is the [pysal](https://pypi.org/project/pysal/) (see [pysal.lib.cg.alpha_shapes](https://www.osgeo.cn/pysal/_modules/pysal/lib/cg/alpha_shapes.html)), the lesser powerful [alphashape](https://pypi.org/project/alphashape/), and the more convoluted [CGAL Python bindings](https://pypi.org/project/cgal-bindings/) and [dionysus](https://pypi.org/project/dionysus/).

### Shapely Code

```python
import matplotlib.pyplot as plt
fig, ax = plt.subplots()

from shapely.ops import polygonize
from shapely.geometry import Polygon
from shapely.geometry import MultiPolygon
from shapely.geometry import MultiLineString
from shapely.geometry import CAP_STYLE, JOIN_STYLE
from shapely.ops import cascaded_union
from shapely.geometry import box
from shapely.geometry import LineString
from shapely.geometry import LinearRing
from shapely.geometry import Point


def plot_coords(coords):
    pts = list(coords)
    x, y = zip(*pts)
    plt.plot(x, y)


def plot_polys(polys):
    for poly in polys:
        if (not getatr(poly, "exterior", None)):
            print("got line?")

        plot_coords(poly.exterior.coords)

        for hole in poly.interiors:
            plot_coords(hole.coords)


class ContainsExteriorPolygon(object):
    def __init__(self, obj):
        self.obj = obj

    def __gt__(self, other):
        return Polygon(self.obj.exterior).contains(other.obj)


if __name__ == "__main__":
    # Four polygons. a big one surrounding them all. two intermediate ones, each with one inside them
    mls = MultiLineString([((0,0),(0,10),(10,10),(10,0),(0,0)),
                        ((5,5),(5,7), (7,7),  (7,5), (5,5)),
                        ((20,0),(20,10),(30,10), (30,0), (20,0)),
                        ((25,5),(25,7), (27,7),  (27,5), (25,5)),
                        ((-50,-50),(-50,50),(50,50),(50,-50),(-50, -50))

    for p in sorted(polygonize(mls), key=ContainsExteriorPolygon, reverse=True):
        plot_polys([p])
        plt.show()
```

**Comment by venugovh** [comment]
Thank you for the suggestions.
